presto-main/src/main/java/com/facebook/presto/operator/scalar/IpPrefixFunctions.java

```diff
@@ -197,6 +197,54 @@ public static Block collapseIpPrefixes(@SqlType("array(IPPREFIX)") Block unsorte
         return blockBuilder.build();
     }
 
+    @Description("Split the input prefix into subnets the size of the new prefix length.")
+    @ScalarFunction("ip_prefix_subnets")
+    @SqlType("array(IPPREFIX)")
+    public static Block ipPrefixSubnets(@SqlType(StandardTypes.IPPREFIX) Slice prefix, @SqlType(StandardTypes.BIGINT) long newPrefixLength)
+    {
+        boolean inputIsIpV4 = isIpv4(prefix);
+
+        if (newPrefixLength < 0 || (inputIsIpV4 && newPrefixLength > 32) || (!inputIsIpV4 && newPrefixLength > 128)) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid prefix length for IPv" + (inputIsIpV4 ? "4" : "6") + ": " + newPrefixLength);
+        }
+
+        int inputPrefixLength = getPrefixLength(prefix);
+        // We can only deaggregate a prefix if the prefix length is longer (more-specific) than the prefix length of the input prefix.
+        // E.g., if the input prefix is /24, the subnetSize must be like /25, /26 etc...
+        if (inputPrefixLength > newPrefixLength) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "newPrefixLength argument '" + newPrefixLength + "' must be longer (more-specific) than the input prefix's length '" + inputPrefixLength + "'");
```

Comment:
would it be possible to just return the original subnet instead?

Reply:
This is a good question. After sampling the behavior in a few other libraries it seems there is no clear answer.
* Throws exception
    * [Python3](https://docs.python.org/3/library/ipaddress.html) and [C#](https://github.com/lduchosal/ipnetwork/blob/master/src/System.Net.IPNetwork/IPNetwork2.cs) subnet functions throw exceptions.
* [netaddr](https://netaddr.readthedocs.io/en/latest/tutorial_01.html#summarizing-list-of-addresses-and-subnets) returns an empty list.
* [ipaddress-go library](https://seancfoley.github.io/IPAddress/) returns the supernet.

I'm in favor of the current behavior of generating a query error because it provides the nice property that all output prefixes will be the same prefix length. It forces the user to determine how their use case should handle the less specific cases which I believe is the best policy for data quality.

Reply:
@elharo @tdcmeehan What's your thoughts on this?

Reply:
I don't have a strong opinion about this. My weak opinion is that compilable SQL functions should not throw errors or otherwise fail at runtime based on the arguments. They should always return a predictable value. Python, C#, and Java are very different things than SQL so I don't find their behavior here a useful precedent.

What happens if input prefix is /24 and subnetsize is also /24?

Reply:
@elharo If the input prefix is /24 and subnetsize is also /24 then we just return the input prefix.

Reply:
The discussion here has stalled a bit. Throwing an error in this case isn't my hill to die on -- I think returning NULL or an empty array could also work. @kaikalur or @feilong-liu, thoughts on this?

Reply:
Thanks to @sharadjaiswal for this suggestion:

> While, I gave this example (for `approx_percentile`), in this case, if someone picks an input percentile value of say 1.35, its really unclear what the `approx_percentile` function should do? Should they treat it as 1.00??  - but that’s inferring the intent of the callee. However, returning ip_prefix_subnets(’[12.13.14.15/24](http://12.13.14.15/24)’, 16) as [] seems more correct - we are not inferring intent, we are just saying “there exist no such subnets”, and it saves us from debugging error messages...